sequencer/src/state.rs

```diff
@@ -45,63 +43,102 @@ impl Default for ValidatedState {
-        let block_merkle_tree_root = block_merkle_tree.commitment();
-        anyhow::ensure!(
-            proposal.block_merkle_tree_root == block_merkle_tree_root,
-            anyhow::anyhow!(
-                "Invalid Block Root Error: {}, {}",
-                block_merkle_tree_root,
-                proposal.block_merkle_tree_root
-            )
-        );
+    let ValidatedState {
+        block_merkle_tree,
+        fee_merkle_tree,
+    } = state;
 
-        // fetch receipts from the l1
-        let receipts = fetch_fee_receipts(parent);
-        for FeeReceipt { recipient, amount } in receipts {
-            // Get the balance in order to add amount, ignoring the proof.
-            match fee_merkle_tree.universal_lookup(recipient) {
-                LookupResult::Ok(balance, _) => fee_merkle_tree
-                    .update(recipient, balance.add(amount))
-                    .unwrap(),
-                // Handle `NotFound` and `NotInMemory` by initializing
-                // state.
-                _ => fee_merkle_tree.update(recipient, amount).unwrap(),
-            };
-        }
+    // validate proposal is descendent of parent by appending to parent
+    block_merkle_tree.push(parent.commit()).unwrap();
+    let block_merkle_tree_root = block_merkle_tree.commitment();
+    anyhow::ensure!(
+        proposal.block_merkle_tree_root == block_merkle_tree_root,
+        anyhow::anyhow!(
+            "Invalid Block Root Error: {}, {}",
+            block_merkle_tree_root,
+            proposal.block_merkle_tree_root
+        )
+    );
 
-        let fee_merkle_tree_root = fee_merkle_tree.commitment();
-        anyhow::ensure!(
-            proposal.fee_merkle_tree_root == fee_merkle_tree_root,
-            anyhow::anyhow!(
-                "Invalid Fee Root Error: {}, {}",
-                fee_merkle_tree_root,
-                proposal.fee_merkle_tree_root
-            )
-        );
+    let fee_merkle_tree_root = fee_merkle_tree.commitment();
+    anyhow::ensure!(
+        proposal.fee_merkle_tree_root == fee_merkle_tree_root,
+        anyhow::anyhow!(
+            "Invalid Fee Root Error: {}, {}",
+            fee_merkle_tree_root,
+            proposal.fee_merkle_tree_root
+        )
+    );
+    Ok(())
+}
 
-        Ok(ValidatedState {
-            block_merkle_tree,
-            fee_merkle_tree,
-        })
+/// Fetch receipts from the l1 and add them to local balance.
+fn update_balance(fee_merkle_tree: &mut FeeMerkleTree, parent: &Header) {
+    let receipts = fetch_fee_receipts(parent);
+    for FeeReceipt { recipient, amount } in receipts {
+        // Get the balance in order to add amount, ignoring the proof.
+        match fee_merkle_tree.universal_lookup(recipient) {
+            LookupResult::Ok(balance, _) => fee_merkle_tree
+                .update(recipient, balance.add(amount))
+                .unwrap(),
+            // Handle `NotFound` and `NotInMemory` by initializing
```

Comment:
`NotInMemory` we will need to handle differently, by looking up the value from a peer. Because it is possible that there is some value for this account, we just don't know it, in which case re-initializing the account is incorrect. Can you just leave a `TODO` here so we remember to fix this when we implement catchup?

Reply:
added in c48d5cbe7f1daf0e9a0dfcb0463a11c67fcc0dfa